File: minstack.py
from collections import deque

class MinStack:
    class Node:
        def __init__(self, val: int, min: int) -> None:
            self.val = val
            self.min = min

    def __init__(self):
        self.stk = deque()

    def push(self, val: int) -> None:
        if len(self.stk) > 0:
            prev_min = self.stk[-1].min
        else:
            prev_min = val
        new_min = min(prev_min, val)
        self.stk.append(self.Node(val, new_min))

    def pop(self) -> None:
        self.stk.pop()

    def top(self) -> int:
        if len(self.stk) > 0:
            return self.stk[-1].val
        return -1

    def getMin(self) -> int:
        return self.stk[-1].min
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: minstack.cpp
#include <vector>
#include <set>
#include <map>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <string>
using namespace std;

class MinStack {
private:
    stack<int> stk1;
    stack<int> stk2;

public:
    MinStack() {}
    
    void push(int val) {
        stk1.push(val);
        if (stk2.size() == 0 || stk2.top() >= val) {
            stk2.push(val);
        }
    }
    
    void pop() {
        int poppingval = stk1.top();
        stk1.pop();
        if (poppingval == stk2.top()) {
            stk2.pop();
        }
    }
    
    int top() {
        return stk1.top();
    }
    
    int getMin() {
        return stk2.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */